include/fastdds/topic/DataReader.hpp

```diff
@@ -0,0 +1,278 @@
+     * Methods to read or take data from the History.
+     */
+
+    ///@{
+
+    bool read(
+            std::vector<void*>& data_values,
+            std::vector<fastrtps::SampleInfo_t>& sample_infos,
+            uint32_t max_samples);
+
+    bool read_next_sample(
+            void* data,
+            fastrtps::SampleInfo_t* info);
+
+    bool take(
+            std::vector<void*>& data_values,
+            std::vector<fastrtps::SampleInfo_t>& sample_infos,
+            uint32_t max_samples);
+
+    bool take_next_sample(
+            void* data,
+            fastrtps::SampleInfo_t* info);
+
+    ///@}
+
+    /**
+    * Get associated GUID
+    * @return Associated GUID
+    */
+    const fastrtps::rtps::GUID_t& guid();
+
+    fastrtps::rtps::InstanceHandle_t get_instance_handle() const;
+
+    /**
+    * Get topic data type
+    * @return Topic data type
+    */
+    TypeSupport type();
+
+    /*!
+    * @brief Returns there is a clean state with all Publishers.
+    * It occurs when the Subscriber received all samples sent by Publishers. In other words,
+    * its WriterProxies are up to date.
+    * @return There is a clean state with all Publishers.
+    */
+    bool is_in_clean_state() const;
+
+    /**
+     * Get the unread count.
+     * @return Unread count
+     */
+    uint64_t get_unread_count() const;
+
+    /**
+     * @brief A method called when a new cache change is added
+     * @param change The cache change that has been added
+     * @return True if the change was added (due to some QoS it could have been 'rejected')
+     */
+    bool on_new_cache_change_added(
+            const fastrtps::rtps::CacheChange_t* const change);
+
+    /**
+     * @brief Get the requested deadline missed status
+     * @return The deadline missed status
+     */
+    void get_requested_deadline_missed_status(
+            fastrtps::RequestedDeadlineMissedStatus& status);
+
+    bool set_attributes(const fastrtps::rtps::ReaderAttributes& att);
+
+    const fastrtps::rtps::ReaderAttributes& get_attributes() const;
+
+    bool set_qos(const fastrtps::ReaderQos& qos);
+
+    const fastrtps::ReaderQos& get_qos() const;
+
+    bool set_topic(const fastrtps::TopicAttributes& att);
+
+    const fastrtps::TopicAttributes& get_topic() const;
+
+    bool set_listener(DataReaderListener* listener);
+
+    const DataReaderListener* get_listener() const;
+
+    bool get_key_value(
+            void* data,
+            const fastrtps::rtps::InstanceHandle_t& handle);
+
+    bool get_liveliness_changed_status(
+            fastrtps::LivelinessChangedStatus& status) const;
+
+    bool get_requested_incompatible_qos_status(
+            fastrtps::RequestedIncompatibleQosStatus& status) const;
+
+    bool get_sample_lost_status(
+            fastrtps::SampleLostStatus& status) const;
+
+    bool get_sample_rejected_status(
+            fastrtps::SampleRejectedStatus& status) const;
+
+    const Subscriber* get_subscriber() const;
```

Comment:
As `Subscriber` and `Publisher` have a `get_participant` method, I though it will be useful for `DataReader` and `DataWriter` to have access to their respective "factories".

Reply:
:+1: